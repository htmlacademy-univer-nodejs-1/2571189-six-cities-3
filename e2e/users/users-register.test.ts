import assert from 'node:assert';
import { Headers, fetch } from 'undici';
import { describe, test } from 'vitest';
import UserRdo from '../../src/modules/user/rdo/user.rdo.js';

process.env['E2E_ENDPOINT'] = 'http://localhost:4000';
assert(process.env['E2E_ENDPOINT'] !== undefined);
assert(process.env['E2E_ENDPOINT'].startsWith('http') === true, 'E2E_ENDPOINT должен начинаться с протокола');

const url = new URL(process.env['E2E_ENDPOINT']);

describe('POST /users/register', async () => {
  test('Success user register', async (tc) => {
    const user = {
      type: 'simple',
      email: 'test@test.com',
      username: 'name',
      password: '12345678910',
    };

    const response = await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    tc.expect(response.ok).toBeTruthy();
    tc.expect(response.status).toStrictEqual(201);
    tc.expect(response.headers.get('content-type')).toMatch(/application\/json/);
    const res = await response.json() as UserRdo;
    tc.expect(res.type).toStrictEqual(user.type);
    tc.expect(res.email).toStrictEqual(user.email);
    tc.expect(res.username).toStrictEqual(user.username);

    const responseWhenUserExist = await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    tc.expect(responseWhenUserExist.ok).toBeFalsy();
    tc.expect(responseWhenUserExist.status).toStrictEqual(409);
    tc.expect(responseWhenUserExist.statusText).toStrictEqual('Conflict');
    tc.expect(responseWhenUserExist.headers.get('content-type')).toMatch(/application\/json/);
    tc.expect(await responseWhenUserExist.json()).toMatchSnapshot();
  });

  test('Invalid email', async (tc) => {
    const user = {
      type: 'simple',
      email: 'invalid',
      username: 'test',
      password: '12345678910',
    };

    const response = await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    tc.expect(response.ok).toBeFalsy();
    tc.expect(response.status).toStrictEqual(400);
    tc.expect(response.statusText).toStrictEqual('Bad Request');
    tc.expect(response.headers.get('content-type')).toMatch(/application\/json/);
    tc.expect(await response.json()).toMatchSnapshot();
  });
});
