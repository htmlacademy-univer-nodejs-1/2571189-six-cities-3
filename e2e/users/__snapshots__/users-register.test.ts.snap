// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`POST /users/register > Invalid email 1`] = `
[
  {
    "children": [],
    "constraints": {
      "isEmail": "Email must be valid.",
    },
    "property": "email",
    "target": {
      "email": "invalid",
      "password": "12345678910",
      "type": "simple",
      "username": "test",
    },
    "value": "invalid",
  },
]
`;

exports[`POST /users/register > Success user register 1`] = `
{
  "error": "User with email test@test.com already exists.",
}
`;
